library/core/src/ptr/mod.rs

```diff
@@ -1135,27 +1135,56 @@ pub const unsafe fn replace<T>(dst: *mut T, mut src: T) -> T {
 #[rustc_const_unstable(feature = "const_ptr_read", issue = "80377")]
 #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
 pub const unsafe fn read<T>(src: *const T) -> T {
-    // We are calling the intrinsics directly to avoid function calls in the generated code
-    // as `intrinsics::copy_nonoverlapping` is a wrapper function.
-    extern "rust-intrinsic" {
-        #[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.63.0")]
-        fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize);
-    }
+    // It would be semantically correct to implement this via `copy_nonoverlapping`
+    // and `MaybeUninit`, as was done before PR #109035.
 
-    let mut tmp = MaybeUninit::<T>::uninit();
-    // SAFETY: the caller must guarantee that `src` is valid for reads.
-    // `src` cannot overlap `tmp` because `tmp` was just allocated on
-    // the stack as a separate allocated object.
+    // However, it switched to intrinsic that lowers to `_0 = *src` in MIR in
+    // order to address a few implementation issues:
     //
-    // Also, since we just wrote a valid value into `tmp`, it is guaranteed
-    // to be properly initialized.
+    // - Using `MaybeUninit::assume_init` after a `copy_nonoverlapping` was not
+    //   turning the untyped copy into a typed load. As such, the generated
+    //   `load` in LLVM didn't get various metadata, such as `!range` (#73258),
+    //   `!nonnull`, and `!noundef`, resulting in poorer optimization.
+    // - Going through the extra local resulted in multiple extra copies, even
+    //   in optimized MIR.  (Ignoring StorageLive/Dead, the intrinsic is one
+    //   MIR statement, while the previous implementation was eight.)  LLVM
```

Comment:
To show my work for these numbers,

With this PR:
```rust
    bb0: {
        StorageLive(_2);                 // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:9:5: 9:23
        _0 = (*_1);                      // scope 2 at C:\src\rust\library\core\src\ptr\mod.rs:1167:13: 1167:50
        StorageDead(_2);                 // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:9:5: 9:23
        return;                          // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:10:2: 10:2
    }
```

Previously:
```rust
    bb0: {
        StorageLive(_6);                 // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:12:5: 12:27
        StorageLive(_2);                 // scope 2 at C:\src\rust\library\core\src\ptr\mod.rs:1195:17: 1195:24
        StorageLive(_7);                 // scope 5 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:314:31: 314:33
        _2 = MaybeUninit::<T> { uninit: move _7 }; // scope 5 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:314:9: 314:35
        StorageDead(_7);                 // scope 5 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:314:34: 314:35
        StorageLive(_3);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:38: 1196:54
        StorageLive(_4);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:38: 1196:54
        _4 = &mut _2;                    // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:38: 1196:54
        StorageLive(_8);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:42: 1196:54
        _8 = &raw mut (*_4);             // scope 6 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:569:9: 569:13
        _3 = _8 as *mut T (PtrToPtr);    // scope 6 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:569:9: 569:33
        StorageDead(_8);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:42: 1196:54
        StorageDead(_4);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:53: 1196:54
        copy_nonoverlapping(dst = move _3, src = _1, count = const 1_usize); // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:13: 1196:58
        StorageDead(_3);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1196:57: 1196:58
        StorageLive(_5);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1197:13: 1197:16
        _5 = move _2;                    // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1197:13: 1197:16
        StorageLive(_9);                 // scope 8 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:627:38: 627:48
        _9 = move (_5.1: std::mem::ManuallyDrop<T>); // scope 8 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:627:38: 627:48
        _0 = move (_9.0: T);             // scope 9 at C:\src\rust\library\core\src\mem\manually_drop.rs:89:9: 89:19
        StorageDead(_9);                 // scope 8 at C:\src\rust\library\core\src\mem\maybe_uninit.rs:627:48: 627:49
        StorageDead(_5);                 // scope 3 at C:\src\rust\library\core\src\ptr\mod.rs:1197:29: 1197:30
        StorageDead(_2);                 // scope 2 at C:\src\rust\library\core\src\ptr\mod.rs:1198:9: 1198:10
        StorageDead(_6);                 // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:12:5: 12:27
        return;                          // scope 0 at C:\src\rust\tests\codegen\aaaaaa-mir-dump-demo.rs:13:2: 13:2
    }
```

That's of course not a semantic problem, as both work fine, but I think it makes a nice demonstration of how this ends up being practically useful, especially in conjunction with MIR inlining.